File: prns-host/bindings/jvm/src/main/kotlin/io/reticulum/prns/Native.kt
package io.reticulum.prns

import com.sun.jna.IntegerType
import com.sun.jna.Library
import com.sun.jna.Memory
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure
import com.sun.jna.ptr.ByteByReference
import com.sun.jna.ptr.LongByReference
import com.sun.jna.ptr.PointerByReference
import java.nio.charset.StandardCharsets

internal class SizeT(value: Long = 0) : IntegerType(Native.SIZE_T_SIZE, value, true) {
    override fun toByte(): Byte = toLong().toByte()

    override fun toShort(): Short = toLong().toShort()
}

@Structure.FieldOrder("data", "length")
internal open class NativeByteView : Structure() {
    @JvmField
    var data: Pointer? = null

    @JvmField
    var length: SizeT = SizeT()

    class ByValue : NativeByteView(), Structure.ByValue

    class ByReference : NativeByteView(), Structure.ByReference
}

@Structure.FieldOrder("data", "length")
internal open class NativeStringView : Structure() {
    @JvmField
    var data: Pointer? = null

    @JvmField
    var length: SizeT = SizeT()

    class ByValue : NativeStringView(), Structure.ByValue
}

@Structure.FieldOrder("structSize", "abi", "schemaVersion", "productVersion")
internal class NativeContractInfo : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var abi: Int = 0

    @JvmField
    var schemaVersion: Int = 0

    @JvmField
    var productVersion: NativeStringView.ByValue = NativeStringView.ByValue()
}

@Structure.FieldOrder(
    "structSize",
    "pendingCommands",
    "applicationEvents",
    "retainedEventBytes",
    "diagnostics",
)
internal open class NativeLimits : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var pendingCommands: SizeT = SizeT()

    @JvmField
    var applicationEvents: SizeT = SizeT()

    @JvmField
    var retainedEventBytes: SizeT = SizeT()

    @JvmField
    var diagnostics: SizeT = SizeT()

    class ByValue : NativeLimits(), Structure.ByValue
}

@Structure.FieldOrder("structSize", "kind", "secret", "path")
internal open class NativeIdentityConfig : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var kind: Int = 0

    @JvmField
    var secret: NativeByteView.ByValue = NativeByteView.ByValue()

    @JvmField
    var path: NativeStringView.ByValue = NativeStringView.ByValue()

    class ByValue : NativeIdentityConfig(), Structure.ByValue
}

@Structure.FieldOrder("structSize", "appName", "aspects", "aspectCount")
internal open class NativeDestinationName : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var appName: NativeStringView.ByValue = NativeStringView.ByValue()

    @JvmField
    var aspects: Pointer? = null

    @JvmField
    var aspectCount: SizeT = SizeT()

    class ByValue : NativeDestinationName(), Structure.ByValue
}

@Structure.FieldOrder(
    "structSize",
    "kind",
    "name",
    "identityKind",
    "dedicatedIdentity",
    "announceAppData",
)
internal open class NativeDestinationConfig : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var kind: Int = 0

    @JvmField
    var name: NativeDestinationName.ByValue = NativeDestinationName.ByValue()

    @JvmField
    var identityKind: Int = 0

    @JvmField
    var dedicatedIdentity: NativeIdentityConfig.ByValue = NativeIdentityConfig.ByValue()

    @JvmField
    var announceAppData: NativeByteView.ByValue = NativeByteView.ByValue()
}

@Structure.FieldOrder(
    "structSize",
    "requiredAbi",
    "requiredProductVersion",
    "limits",
    "role",
    "identity",
    "destinations",
    "destinationCount",
    "requiredCapabilities",
    "requiredCapabilityCount",
)
internal class NativeHostOptions : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var requiredAbi: Int = 0

    @JvmField
    var requiredProductVersion: NativeStringView.ByValue = NativeStringView.ByValue()

    @JvmField
    var limits: NativeLimits.ByValue = NativeLimits.ByValue()

    @JvmField
    var role: Int = 0

    @JvmField
    var identity: NativeIdentityConfig.ByValue = NativeIdentityConfig.ByValue()

    @JvmField
    var destinations: Pointer? = null

    @JvmField
    var destinationCount: SizeT = SizeT()

    @JvmField
    var requiredCapabilities: Pointer? = null

    @JvmField
    var requiredCapabilityCount: SizeT = SizeT()
}

@Structure.FieldOrder("structSize", "outcome", "failure", "evidence", "rttMillis", "value", "detail")
internal class NativeCommandResult : Structure() {
    @JvmField
    var structSize: SizeT = SizeT()

    @JvmField
    var outcome: Int = 0

    @JvmField
    var failure: Int = 0

    @JvmField
    var evidence: Int = 0

    @JvmField
    var rttMillis: Long = 0

    @JvmField
    var value: NativeByteView.ByValue = NativeByteView.ByValue()

    @JvmField
    var detail: NativeStringView.ByValue = NativeStringView.ByValue()
}

internal interface PrnsNative : Library {
    fun prns_contract_info(info: NativeContractInfo): Int
    fun prns_host_create(options: NativeHostOptions, host: PointerByReference): Int
    fun prns_host_release(host: Pointer)
    fun prns_host_identity_hash(host: Pointer, hash: NativeByteView): Int
    fun prns_host_destination_count(host: Pointer): SizeT
    fun prns_host_destination_hash(host: Pointer, index: SizeT, hash: NativeByteView): Int
    fun prns_host_announce(
        host: Pointer,
        destination: NativeByteView.ByValue,
        interfaceId: NativeByteView.ByReference?,
        command: PointerByReference,
    ): Int
    fun prns_host_send_single_packet(
        host: Pointer,
        destination: NativeByteView.ByValue,
        payload: NativeByteView.ByValue,
        command: PointerByReference,
    ): Int
    fun prns_host_close_link(
        host: Pointer,
        linkId: NativeByteView.ByValue,
        command: PointerByReference,
    ): Int
    fun prns_host_attach_tcp_server(
        host: Pointer,
        bind: NativeStringView.ByValue,
        bitrateKind: Int,
        bitrateBps: Long,
        command: PointerByReference,
    ): Int
    fun prns_host_attach_tcp_client(
        host: Pointer,
        target: NativeStringView.ByValue,
        bitrateKind: Int,
        bitrateBps: Long,
        command: PointerByReference,
    ): Int
    fun prns_host_attach_udp(
        host: Pointer,
        local: NativeStringView.ByValue,
        peer: NativeStringView.ByValue,
        bitrateKind: Int,
        bitrateBps: Long,
        command: PointerByReference,
    ): Int
    fun prns_host_detach_interface(
        host: Pointer,
        interfaceId: NativeByteView.ByValue,
        command: PointerByReference,
    ): Int
    fun prns_host_stop(host: Pointer): Int
    fun prns_command_wait(command: Pointer, timeoutMillis: Int, result: NativeCommandResult): Int
    fun prns_command_interrupt_wait(command: Pointer)
    fun prns_command_release(command: Pointer)
    fun prns_host_claim_application_events(host: Pointer, stream: PointerByReference): Int
    fun prns_host_claim_diagnostics(host: Pointer, stream: PointerByReference): Int
    fun prns_event_stream_interrupt_wait(stream: Pointer)
    fun prns_event_stream_release(stream: Pointer)
    fun prns_event_stream_next(
        stream: Pointer,
        timeoutMillis: Int,
        event: PointerByReference,
    ): Int
    fun prns_event_release(event: Pointer)
    fun prns_event_kind(event: Pointer): Int
    fun prns_event_bytes(event: Pointer, field: Int, value: NativeByteView): Int
    fun prns_event_string(event: Pointer, field: Int, value: NativeStringView): Int
    fun prns_event_u64(event: Pointer, field: Int, value: LongByReference): Int
    fun prns_event_u128(
        event: Pointer,
        field: Int,
        low: LongByReference,
        high: LongByReference,
    ): Int
    fun prns_event_resource_stream(event: Pointer, stream: PointerByReference): Int
    fun prns_resource_stream_release(stream: Pointer)
    fun prns_resource_stream_next(
        stream: Pointer,
        maximumBytes: SizeT,
        chunk: NativeByteView,
        finished: ByteByReference,
    ): Int
}

internal object NativeApi {
    val library: PrnsNative by lazy {
        val path = System.getProperty("personal.rns.library")
            ?: System.getenv("PRNS_HOST_LIBRARY")
            ?: "prns_host"
        Native.load(path, PrnsNative::class.java)
    }
}

class StatusException(
    val operation: String,
    val status: Status,
) : RuntimeException("$operation failed with $status")

internal fun checkedStatus(rawValue: Int, operation: String) {
    val status = Status.fromRawValue(rawValue) ?: Status.BACKEND_FAILED
    if (status != Status.OK) {
        throw StatusException(operation, status)
    }
}

internal fun copyBytes(view: NativeByteView): ByteArray {
    val length = view.length.toLong()
    if (length == 0L) {
        return ByteArray(0)
    }
    return requireNotNull(view.data).getByteArray(0, length.toInt())
}

internal fun copyString(view: NativeStringView): String {
    val length = view.length.toLong()
    if (length == 0L) {
        return ""
    }
    return String(
        requireNotNull(view.data).getByteArray(0, length.toInt()),
        StandardCharsets.UTF_8,
    )
}

internal class NativeArena : AutoCloseable {
    private val allocations = mutableListOf<Memory>()
    private val structures = mutableListOf<Structure>()

    fun bytes(value: ByteArray): NativeByteView.ByValue {
        val result = NativeByteView.ByValue()
        if (value.isNotEmpty()) {
            val memory = Memory(value.size.toLong())
            memory.write(0, value, 0, value.size)
            allocations += memory
            result.data = memory
            result.length = SizeT(value.size.toLong())
        }
        result.write()
        return result
    }

    fun bytesReference(value: ByteArray): NativeByteView.ByReference {
        val view = bytes(value)
        return NativeByteView.ByReference().also {
            it.data = view.data
            it.length = view.length
            it.write()
            structures += it
        }
    }

    fun string(value: String): NativeStringView.ByValue {
        val bytes = value.toByteArray(StandardCharsets.UTF_8)
        val result = NativeStringView.ByValue()
        if (bytes.isNotEmpty()) {
            val memory = Memory(bytes.size.toLong())
            memory.write(0, bytes, 0, bytes.size)
            allocations += memory
            result.data = memory
            result.length = SizeT(bytes.size.toLong())
        }
        result.write()
        return result
    }

    fun identity(value: IdentityConfig): NativeIdentityConfig.ByValue {
        val result = NativeIdentityConfig.ByValue()
        result.structSize = SizeT(result.size().toLong())
        when (value) {
            is IdentityConfigExisting -> {
                val secret = value.secret.copyBytes()
                try {
                    result.kind = IdentityConfigKind.EXISTING.rawValue
                    result.secret = bytes(secret)
                } finally {
                    secret.fill(0)
                }
            }
            IdentityConfigGenerateEphemeral -> {
                result.kind = IdentityConfigKind.GENERATE_EPHEMERAL.rawValue
            }
            is IdentityConfigLoadOrCreate -> {
                result.kind = IdentityConfigKind.LOAD_OR_CREATE.rawValue
                result.path = string(value.path)
            }
        }
        result.write()
        return result
    }

    private fun destinationName(value: DestinationName): NativeDestinationName.ByValue {
        val result = NativeDestinationName.ByValue()
        result.structSize = SizeT(result.size().toLong())
        result.appName = string(value.appName)
        if (value.aspects.isNotEmpty()) {
            val first = NativeStringView()
            val array = first.toArray(value.aspects.size)
            value.aspects.forEachIndexed { index, aspect ->
                val item = array[index] as NativeStringView
                val native = string(aspect)
                item.data = native.data
                item.length = native.length
                item.write()
                structures += item
            }
            result.aspects = first.pointer
            result.aspectCount = SizeT(value.aspects.size.toLong())
            structures += first
        }
        result.write()
        return result
    }

    private fun destination(value: DestinationConfig): NativeDestinationConfig {
        val result = NativeDestinationConfig()
        result.structSize = SizeT(result.size().toLong())
        when (value) {
            is DestinationConfigPlain -> {
                result.kind = DestinationConfigKind.PLAIN.rawValue
                result.name = destinationName(value.name)
            }
            is DestinationConfigSingle -> {
                result.kind = DestinationConfigKind.SINGLE.rawValue
                result.name = destinationName(value.name)
                when (val identity = value.identity) {
                    DestinationIdentityConfigHostIdentity -> {
                        result.identityKind =
                            DestinationIdentityConfigKind.HOST_IDENTITY.rawValue
                    }
                    is DestinationIdentityConfigDedicatedIdentity -> {
                        result.identityKind =
                            DestinationIdentityConfigKind.DEDICATED_IDENTITY.rawValue
                        result.dedicatedIdentity = identity(identity.identity)
                    }
                }
                value.announceAppData?.let {
                    result.announceAppData = bytes(it.copyBytes())
                }
            }
        }
        result.write()
        return result
    }

    fun hostOptions(value: HostOptions): NativeHostOptions {
        val result = NativeHostOptions()
        result.structSize = SizeT(result.size().toLong())
        result.requiredAbi = HostContract.ABI
        result.requiredProductVersion = string(HostContract.PRODUCT_VERSION)
        result.limits = NativeLimits.ByValue().also {
            it.structSize = SizeT(it.size().toLong())
            it.pendingCommands = SizeT(value.limits.pendingCommands)
            it.applicationEvents = SizeT(value.limits.applicationEvents)
            it.retainedEventBytes = SizeT(value.limits.retainedEventBytes)
            it.diagnostics = SizeT(value.limits.diagnostics)
            it.write()
        }
        result.role = value.role.rawValue
        result.identity = identity(value.identity)
        if (value.destinations.isNotEmpty()) {
            val first = NativeDestinationConfig()
            val array = first.toArray(value.destinations.size)
            value.destinations.forEachIndexed { index, destination ->
                val target = array[index] as NativeDestinationConfig
                val native = destination(destination)
                target.structSize = native.structSize
                target.kind = native.kind
                target.name = native.name
                target.identityKind = native.identityKind
                target.dedicatedIdentity = native.dedicatedIdentity
                target.announceAppData = native.announceAppData
                target.write()
                structures += target
            }
            result.destinations = first.pointer
            result.destinationCount = SizeT(value.destinations.size.toLong())
            structures += first
        }
        if (value.requiredCapabilities.isNotEmpty()) {
            val capabilities = value.requiredCapabilities.sortedBy(Capability::rawValue)
            val memory = Memory(Int.SIZE_BYTES.toLong() * capabilities.size)
            capabilities.forEachIndexed { index, capability ->
                memory.setInt(Int.SIZE_BYTES.toLong() * index, capability.rawValue)
            }
            allocations += memory
            result.requiredCapabilities = memory
            result.requiredCapabilityCount = SizeT(capabilities.size.toLong())
        }
        result.write()
        return result
    }

    override fun close() {
        allocations.asReversed().forEach {
            it.clear()
            it.close()
        }
        structures.clear()
        allocations.clear()
    }
}

internal fun verifyNativeContract() {
    val info = NativeContractInfo()
    info.structSize = SizeT(info.size().toLong())
    info.write()
    checkedStatus(NativeApi.library.prns_contract_info(info), "contractInfo")
    info.read()
    if (
        info.abi != HostContract.ABI ||
        info.schemaVersion != HostContract.SCHEMA_VERSION ||
        copyString(info.productVersion) != HostContract.PRODUCT_VERSION
    ) {
        throw StatusException("contractInfo", Status.CONTRACT_MISMATCH)
    }
}

File: prns-host/bindings/julia/src/native.jl
const NEVER_TIMEOUT = typemax(UInt32)
const EVENT_WAIT_SLICE_MILLIS = UInt32(50)
const NATIVE_LIBRARY = Ref{Ptr{Cvoid}}(C_NULL)

struct NativeByteView
    data::Ptr{UInt8}
    length::Csize_t
end

const NativeStringView = NativeByteView

struct NativeContractInfo
    struct_size::Csize_t
    abi::UInt32
    schema_version::UInt32
    product_version::NativeStringView
end

struct NativeLimits
    struct_size::Csize_t
    pending_commands::Csize_t
    application_events::Csize_t
    retained_event_bytes::Csize_t
    diagnostics::Csize_t
end

struct NativeIdentityConfig
    struct_size::Csize_t
    kind::UInt32
    secret::NativeByteView
    path::NativeStringView
end

struct NativeDestinationName
    struct_size::Csize_t
    app_name::NativeStringView
    aspects::Ptr{NativeStringView}
    aspect_count::Csize_t
end

struct NativeRequestHandlerConfig
    struct_size::Csize_t
    path::NativeStringView
    policy::UInt32
end

struct NativeDestinationConfig
    struct_size::Csize_t
    kind::UInt32
    name::NativeDestinationName
    identity_kind::UInt32
    dedicated_identity::NativeIdentityConfig
    announce_app_data::NativeByteView
    request_handlers::Ptr{NativeRequestHandlerConfig}
    request_handler_count::Csize_t
end

struct NativeHostOptions
    struct_size::Csize_t
    required_abi::UInt32
    required_product_version::NativeStringView
    limits::NativeLimits
    role::UInt32
    identity::NativeIdentityConfig
    destinations::Ptr{NativeDestinationConfig}
    destination_count::Csize_t
    required_capabilities::Ptr{UInt32}
    required_capability_count::Csize_t
end

struct NativeCommandResult
    struct_size::Csize_t
    outcome::UInt32
    failure::UInt32
    evidence::UInt32
    rtt_millis::UInt64
    value::NativeByteView
    detail::NativeStringView
end

mutable struct NativeArena
    buffers::Vector{Vector{UInt8}}
    string_arrays::Vector{Vector{NativeStringView}}
    request_handler_arrays::Vector{Vector{NativeRequestHandlerConfig}}
end

NativeArena() = NativeArena(
    Vector{UInt8}[],
    Vector{NativeStringView}[],
    Vector{NativeRequestHandlerConfig}[],
)

function Base.close(arena::NativeArena)
    foreach(buffer -> fill!(buffer, 0), arena.buffers)
    empty!(arena.buffers)
    empty!(arena.string_arrays)
    empty!(arena.request_handler_arrays)
    nothing
end

function artifact_native_library()
    artifacts_toml = normpath(joinpath(@__DIR__, "..", "Artifacts.toml"))
    isfile(artifacts_toml) || return ""
    PkgArtifacts.ensure_artifact_installed(
        "personal_rns",
        artifacts_toml,
    )
    hash = Artifacts.artifact_hash("personal_rns", artifacts_toml)
    hash === nothing && return ""
    artifact = Artifacts.artifact_path(hash)
    roots = [artifact]
    append!(
        roots,
        filter(isdir, readdir(artifact; join=true)),
    )
    for root in roots
        for name in ("libprns_host.so", "libprns_host.dylib", "prns_host.dll")
            candidate = joinpath(root, "lib", name)
            isfile(candidate) && return candidate
        end
    end
    ""
end

function native_library()
    NATIVE_LIBRARY[] != C_NULL && return NATIVE_LIBRARY[]
    configured = get(ENV, "PRNS_HOST_LIBRARY", "")
    candidate = configured
    isempty(candidate) && (candidate = artifact_native_library())
    isempty(candidate) &&
        (candidate = Libdl.find_library(["prns_host", "libprns_host"]))
    isempty(candidate) && throw(NativeLibraryUnavailable())
    NATIVE_LIBRARY[] = Libdl.dlopen(candidate)
    NATIVE_LIBRARY[]
end

native_symbol(name::Symbol) = Libdl.dlsym(native_library(), name)

struct NativeLibraryUnavailable <: Exception end

Base.showerror(io::IO, ::NativeLibraryUnavailable) =
    print(io, "Personal RNS native library is unavailable")

struct StatusFailure <: Exception
    operation::Symbol
    status::Status
end

function Base.showerror(io::IO, failure::StatusFailure)
    print(io, "Personal RNS ", failure.operation, " failed with ", failure.status)
end

function checked_status(operation::Symbol, raw::UInt32)
    status = Status(raw)
    status == StatusOk || throw(StatusFailure(operation, status))
    status
end

function native_byte_view(arena::NativeArena, value)
    bytes = UInt8[item for item in value]
    push!(arena.buffers, bytes)
    data_pointer = isempty(bytes) ? Ptr{UInt8}(C_NULL) : pointer(bytes)
    NativeByteView(data_pointer, length(bytes))
end

function native_optional_byte_view(
    arena::NativeArena,
    value::Union{Nothing,AbstractVector{UInt8}},
)
    value === nothing && return NativeByteView(C_NULL, 0)
    bytes = isempty(value) ? UInt8[0] : Vector{UInt8}(value)
    push!(arena.buffers, bytes)
    NativeByteView(pointer(bytes), length(value))
end

function native_string_view(arena::NativeArena, value::AbstractString)
    native_byte_view(arena, Vector{UInt8}(codeunits(value)))
end

function copy_view(view::NativeByteView)
    view.length == 0 && return UInt8[]
    unsafe_wrap(Vector{UInt8}, view.data, Int(view.length)) |> copy
end

copy_string(view::NativeStringView) = String(copy_view(view))

function native_identity(arena::NativeArena, value::IdentityConfig)
    if value isa IdentityConfigExisting
        return NativeIdentityConfig(
            sizeof(NativeIdentityConfig),
            UInt32(IdentityConfigKindExisting),
            native_byte_view(arena, value.secret.bytes),
            NativeStringView(C_NULL, 0),
        )
    end
    if value isa IdentityConfigGenerateEphemeral
        return NativeIdentityConfig(
            sizeof(NativeIdentityConfig),
            UInt32(IdentityConfigKindGenerateEphemeral),
            NativeByteView(C_NULL, 0),
            NativeStringView(C_NULL, 0),
        )
    end
    if value isa IdentityConfigLoadOrCreate
        return NativeIdentityConfig(
            sizeof(NativeIdentityConfig),
            UInt32(IdentityConfigKindLoadOrCreate),
            NativeByteView(C_NULL, 0),
            native_string_view(arena, value.path),
        )
    end
    throw(ArgumentError("unknown identity configuration"))
end

function native_destination_name(arena::NativeArena, value::DestinationName)
    aspects = NativeStringView[
        native_string_view(arena, aspect) for aspect in value.aspects
    ]
    push!(arena.string_arrays, aspects)
    NativeDestinationName(
        sizeof(NativeDestinationName),
        native_string_view(arena, value.app_name),
        isempty(aspects) ? C_NULL : pointer(aspects),
        length(aspects),
    )
end

function native_destination_identity(
    arena::NativeArena,
    value::DestinationIdentityConfig,
)
    if value isa DestinationIdentityConfigHostIdentity
        return (
            UInt32(DestinationIdentityConfigKindHostIdentity),
            NativeIdentityConfig(
                sizeof(NativeIdentityConfig),
                0,
                NativeByteView(C_NULL, 0),
                NativeStringView(C_NULL, 0),
            ),
        )
    end
    if value isa DestinationIdentityConfigDedicatedIdentity
        return (
            UInt32(DestinationIdentityConfigKindDedicatedIdentity),
            native_identity(arena, value.identity),
        )
    end
    throw(ArgumentError("unknown destination identity configuration"))
end

function native_destination(arena::NativeArena, value::DestinationConfig)
    if value isa DestinationConfigPlain
        return NativeDestinationConfig(
            sizeof(NativeDestinationConfig),
            UInt32(DestinationConfigKindPlain),
            native_destination_name(arena, value.name),
            0,
            NativeIdentityConfig(
                sizeof(NativeIdentityConfig),
                0,
                NativeByteView(C_NULL, 0),
                NativeStringView(C_NULL, 0),
            ),
            NativeByteView(C_NULL, 0),
            C_NULL,
            0,
        )
    end
    if value isa DestinationConfigSingle
        identity_kind, identity = native_destination_identity(
            arena,
            value.identity,
        )
        request_handlers = NativeRequestHandlerConfig[
            NativeRequestHandlerConfig(
                sizeof(NativeRequestHandlerConfig),
                native_string_view(arena, handler.path),
                UInt32(handler.policy),
            )
            for handler in value.request_handlers
        ]
        push!(arena.request_handler_arrays, request_handlers)
        return NativeDestinationConfig(
            sizeof(NativeDestinationConfig),
            UInt32(DestinationConfigKindSingle),
            native_destination_name(arena, value.name),
            identity_kind,
            identity,
            native_optional_byte_view(arena, value.announce_app_data),
            isempty(request_handlers) ? C_NULL : pointer(request_handlers),
            length(request_handlers),
        )
    end
    throw(ArgumentError("unknown destination configuration"))
end

function native_host_options(arena::NativeArena, value)
    destinations = NativeDestinationConfig[
        native_destination(arena, destination)
        for destination in value.destinations
    ]
    capabilities = UInt32[
        UInt32(capability) for capability in value.required_capabilities
    ]
    NativeHostOptions(
        sizeof(NativeHostOptions),
        HOST_CONTRACT_ABI,
        native_string_view(arena, PRODUCT_VERSION),
        NativeLimits(
            sizeof(NativeLimits),
            value.limits.pending_commands,
            value.limits.application_events,
            value.limits.retained_event_bytes,
            value.limits.diagnostics,
        ),
        UInt32(value.role),
        native_identity(arena, value.identity),
        isempty(destinations) ? C_NULL : pointer(destinations),
        length(destinations),
        isempty(capabilities) ? C_NULL : pointer(capabilities),
        length(capabilities),
    ), destinations, capabilities
end

function verify_contract()
    output = Ref(
        NativeContractInfo(
            sizeof(NativeContractInfo),
            0,
            0,
            NativeStringView(C_NULL, 0),
        ),
    )
    checked_status(
        :contract_info,
        ccall(
            native_symbol(:prns_contract_info),
            UInt32,
            (Ref{NativeContractInfo},),
            output,
        ),
    )
    actual = output[]
    actual.abi == HOST_CONTRACT_ABI ||
        throw(StatusFailure(:contract_info, StatusContractMismatch))
    actual.schema_version == HOST_SCHEMA_VERSION ||
        throw(StatusFailure(:contract_info, StatusContractMismatch))
    copy_string(actual.product_version) == PRODUCT_VERSION ||
        throw(StatusFailure(:contract_info, StatusContractMismatch))
    nothing
end
